app/src/shared/hooks/useBalance.js

```diff
@@ -0,0 +1,69 @@
+// @flow
+
+import { useState, useEffect, useCallback, useMemo } from 'react'
+
+import { getDataTokenBalance, getEthBalance } from '$mp/utils/web3'
+import type { Address } from '$shared/flowtype/web3-types'
+
+export const BalanceType = {
+    ETH: 'ETH',
+    DATA: 'DATA',
+}
+
+export async function getBalance(address: Address, balanceType: $Values<typeof BalanceType>, usePublicNode?: boolean = false) {
+    let balance
+
+    if (balanceType === BalanceType.ETH) {
+        balance = await getEthBalance(address, usePublicNode)
+    } else if (balanceType === BalanceType.DATA) {
+        balance = await getDataTokenBalance(address, usePublicNode)
+    } else {
+        throw new Error('Unknown balance type!')
+    }
+
+    return balance
+}
+
+export function useBalance(account: Address, balanceType: $Values<typeof BalanceType>) {
+    const [fetching, setFetching] = useState(false)
+    const [balance, setBalance] = useState(undefined)
+    const [error, setError] = useState(undefined)
+
+    const loadBalance = useCallback(async (address, type) => {
+        setFetching(true)
+        setBalance(undefined)
+
+        try {
+            const newBalance = await getBalance(address, type, true)
+
```

Comment:
check for unmount? Also, if address/type changes before this resolves, then could `setBalance` end up setting the balance for the wrong address/type?

Reply:
It's a good idea but what would be the best way to do that? Store balances to state as an object with `address` and `type` key combination?